### PR TITLE
CTPPS: diamond mapping for 2017 data taking

### PR DESCRIPTION
**Pull request by forthommel**
This PR introduces the channels mapping for 2017 diamond detectors operations, along with a limited set of small changes:
* the `CTPPSDiamondDetId` object is modified to account for a "more trivial" list of channels.
* the CRC validity check for VFAT frames is directly disabled (along with the events counter check) in the `ctppsDiamondRawToDigi` object definition.

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
On 4/27/17 1:42 AM, Laurent Forthomme wrote:
> This PR introduces the channels mapping for 2017 diamond detectors
> operations, along with a limited set of small changes:
> 

is this required for data commissioning in May?
How early in May?

**Comment by forthommel** [comment]
Hi @slava77 
It would indeed be required for the alignment runs foreseen around the ~26th May. But of course, as usual, the sooner the better!

**Comment by civanch** [comment]
+1

**Comment by perrotta** [comment]
+1
Changes to code and configs are correct.
Successfully tested with 2016H data with the test configs available in EventFilter/CTPPSRawToDigi/test
I assume the mapping for 2017 is the correct one, as well as the validity range for the new settings.

**Comment by slava77** [comment]
On 4/28/17 3:30 AM, perrotta wrote:
> +1
> Changes to code and configs are correct.
> Successfully tested with 2016H data with the test configs available in
> EventFilter/CTPPSRawToDigi/test
> I assume the mapping for 2017 is the correct one, as well as the
> validity range for the new settings.

Is there any local run data available from 2017 to test new changes?

**Comment by forthommel** [comment]
Hi @slava77 
You may use the miniDAQ runs 292529, and [the following ones](https://cmswbm.cern.ch/cmsdb/servlet/RunSummary?DB=default&RUN_BEGIN=292529&RUN_END=295000&STATUS_CTPPS_TOT=on&SUBMIT_BOTTOM=SubmitQuery) still present on the Tier-2 (e.g. `/store/t0streamer/Minidaq/A/000/292/529/`)

**Comment by perrotta** [comment]
@forthommel : in /t0streamer/Minidaq/A/000/292/529/ there are no root files to play with.

The concern of @slava77 was that there could be some mistake in the mapping as implemented in the xml file (as all other aspects of this PR got tested): could you please run yourself the test config on a  2017 data file, look at the content of the relevant collections produced, and report here about the correctness of that mapping? (Or, alternatively, point me to a root file with data, or copy it in some public afs area so that I can process it). 

Thx

**Comment by perrotta** [comment]
hold
(until the mapping for 2017 data taking get certified)

**Comment by forthommel** [comment]
Hi @perrotta 
Indeed, the files you can find in this directory are the raw (binary) streamer outputs from miniDAQ. You may use a `NewEventStreamFileReader` source to access them, or use 
[this example configuration file](https://github.com/cms-sw/cmssw/files/965992/ctpps_diamond_mapping_test_cfg.py.txt) and look at the `ctppsDiamondRawToDigi` collection inside (no rechits nor local tracks are expected, as these runs are produced without any beam... fortunately!)
Or if you prefer I can produce a raw EDM file out of these `t0streamer` binaries?

**Comment by perrotta** [comment]
@forthommel : can you please have a look yourself? You know better than me where exactly to look at in order to check that the whole new mapping in the xml file of this PR is correct. Then you can post here that evidence (either plot or text output). Thank you!

**Comment by slava77** [comment]
On 4/29/17 1:08 AM, Laurent Forthomme wrote:
> Hi @perrotta <https://github.com/perrotta>
> Indeed, the files you can find in this directory are the raw (binary)
> streamer outputs from miniDAQ. You may use a |NewEventStreamFileReader|
> source to access them, or use
> this example configuration file
> <https://github.com/cms-sw/cmssw/files/965992/ctpps_diamond_mapping_test_cfg.py.txt>
> and look at the |ctppsDiamondRawToDigi| collection inside (no rechits
> nor local tracks are expected, as these runs are produced without any
> beam... fortunately!)
> Or if you prefer I can produce a raw EDM file out of these |t0streamer|
> binaries?
> 

are there any files with charge injection?
so that there are hit payloads in the data frames.

If readout is not zero suppressed, then existing files are OK.

please clarify

Thank you.

**Comment by forthommel** [comment]
Hi @slava77, @perrotta,

Forwarding @nminafra's comment: we have a complex (and analog) cabling, hence the only way to check the mapping is using the strips and the beam. We have a set of "tomography plots" in our diamond DQM for this.
This cabling was carefully done, but unfortunately we have no way to cross check without the beam. However, this will not produce any error or warning, only the tracks will be affected.
But again all the digital part has been checked (incl. the channel id). It is assigned by the digitizer board in the tunnel and is also written in the xml.
If an id does not correspond to the correct position in the frame, the unpacker should raise an error (and it is not happening in our tests).

Bottom line: we checked and double checked the mapping, in case of error it will not cause warnings or errors but, unfortunately, we could have some misplaced channels that will require a revised version (only) of the xml after the first data taking during the alignment run. 

As for your last comment, @slava77, we unfortunately do not have any charge injection mechanism on the HW side (but dark counts are negligibles), and zero suppression is not yet implemented in our FW.

**Comment by slava77** [comment]
On 4/29/17 8:01 AM, Laurent Forthomme wrote:
> Hi @slava77 <https://github.com/slava77>, @perrotta
> <https://github.com/perrotta>,
> 
> Forwarding @nminafra <https://github.com/nminafra>'s comment: we have a
> complex (and analog) cabling, hence the only way to check the mapping is
> using the strips and the beam. We have a set of "tomography plots" in
> our diamond DQM for this.
> This cabling was carefully done, but unfortunately we have no way to
> cross check without the beam. However, this will not produce any error
> or warning, only the tracks will be affected.
> But again all the digital part has been checked (incl. the channel id).
> It is assigned by the digitizer board in the tunnel and is also written
> in the xml.
> If an id does not correspond to the correct position in the frame, the
> unpacker should raise an error (and it is not happening in our tests).
> 
> Bottom line: we checked and double checked the mapping, in case of error
> it will not cause warnings or errors but, unfortunately, we could have
> some misplaced channels that will require a revised version (only) of
> the xml after the first data taking during the alignment run.
> 
> As for your last comment, @slava77 <https://github.com/slava77>, we
> unfortunately do not have any charge injection mechanism on the HW side
> (but dark counts are negligibles), and zero suppression is not yet
> implemented in our FW.

OK.

Just to be sure, have your tests on 2017 data been done with the code
directly from the PR?
If correctness of all logic can't be checked due to lack of non-zero inputs,
we should at least be sure that all relevant files are committed and run
OK even on these zeroed-out inputs.

**Comment by forthommel** [comment]
Thanks for this fast reaction! Yes, these tests were performed on top of this PR.

**Comment by forthommel** [comment]
Hi @slava77, @perrotta,
Just to be sure, is there any action needed on our side to proceed with this PR?

**Comment by perrotta** [comment]
@forthommel : I'm already fine with it. If also @slava77 is ok with your answers I will un-hold

**Comment by slava77** [comment]
On 5/2/17 2:00 AM, perrotta wrote:
> @forthommel <https://github.com/forthommel> : I'm already fine with it.
> If also @slava77 <https://github.com/slava77> is ok with your answers I
> will un-hold
> 

fine with me

**Comment by perrotta** [comment]
unhold

File: CondFormats/CTPPSReadoutObjects/xml/mapping_timing_diamond_2017.xml [hunk 1]
```diff
@@ -0,0 +1,129 @@
+<top>
```
Comment (franzoni) [inline CondFormats/CTPPSReadoutObjects/xml/mapping_timing_diamond_2017.xml]: hello @forthommel 

are you planning on moving this mapping content to the conditions database ?

I guess you need CTPPSPixelDAQMappingRcd 
to be provided to you from this parallel pr https://github.com/cms-sw/cmssw/pull/18359/files#diff-03e8ca1cea42e0df6e5deb53fc5dd4be ?

Comment (arunhep) [inline CondFormats/CTPPSReadoutObjects/xml/mapping_timing_diamond_2017.xml]: we're looking at https://github.com/cms-sw/cmssw/pull/18359 in parallel

**Comment by franzoni** [comment]
in fact 
https://github.com/cms-sw/cmssw/pull/18359
provides tools to turn release-bound CTPPS conditions
into sqlite , which can be trivially loaded into offline Database

particularly if you anticipate,
as per explanations earlier in this thread,
that corrections to the mapping might become available as soon as you'll be able to validate with real data,
integrating the mapping in a global tag
will ease/speedup deploying such corrections

This needs a focussed discussion on tag integration and GT update
which is beyond the scope of this chat
( please contact AlCa/Db , please be invoited to the Monday 8th AlCa/Db meeting )

**Comment by franzoni** [comment]
+1

this PR is ok
we trust on a follow up to the operational questions of above

**Comment by ggovi** [comment]
+1

**Comment by forthommel** [comment]
Many thanks for your reviews!
Please notice that this PR concerns the diamond detectors mapping, and not the pixels detectors (as covered by the PR you mentioned).
Of course, as soon as the manpower becomes available on our side we will indeed integrate this eras-based mapping into the frontier database!

File: CondFormats/CTPPSReadoutObjects/xml/mapping_timing_diamond_2017.xml (hunk 1 above)
Comment (forthommel) [inline CondFormats/CTPPSReadoutObjects/xml/mapping_timing_diamond_2017.xml]: Hi @franzoni 
As noted in https://github.com/cms-sw/cmssw/pull/18490#issuecomment-298711098, this PR only tackles the diamond detectors part (and not the pixel detectors using this `CTPPSPixelDAQMappingRcd` record). Hence, a follow-up is indeed expected for the migration of this xml mapping to the conditions db!

**Comment by davidlange6** [comment]
+1
